### PR TITLE
feat: add duckdb consumer and tpch tests

### PR DESCRIPTION
**Pull request by richtia**
PR to add tpch query tests using DuckDB as a consumer

### CODE REVIEW

File: tests/integration/test_duckdb_tpch.py
Comment (pdet) [inline tests/integration/test_duckdb_tpch.py:74]: It seems this function is DuckDB consuming its own produced substrait plans, right?

If that's the case (and goal), you can simplify the code by setting `PRAGMA enable_verification`.

Basically, when that pragma is set, DuckDB will internally execute the query and the substrait plan resulting from `get_substrait(SQL)` and compare it's resulting throwing an Error if they do not match.

I think the verify_equals is more interesting if you are comparing results of two different systems consuming a substrait plan.

File: tests/integration/test_duckdb_tpch.py
Comment (pdet) [inline tests/integration/test_duckdb_tpch.py:91]: Just to clarify, I do this sorting on column names, because internally we do test the consumption of ibis substrait plans vs ibis query results, and ibis does not output the results in the order set by the SQL query. Hence, this does not seem to be necessary here.

File: tests/consumers/duckdb_consumer.py
Comment (pdet) [inline tests/consumers/duckdb_consumer.py:22]: What does json formatted string representing the substrait query plan means in this context? Isn't this just a plain SQL query string?

File: tests/integration/test_duckdb_tpch.py
Comment (richtia) [reply tests/integration/test_duckdb_tpch.py:74]: > It seems this function is DuckDB consuming its own produced substrait plans, right?
> 
> If that's the case (and goal), you can simplify the code by setting `PRAGMA enable_verification`.
> 
> Basically, when that pragma is set, DuckDB will internally execute the query and the substrait plan resulting from `get_substrait(SQL)` and compare it's resulting throwing an Error if they do not match.
> 
> I think the verify_equals is more interesting if you are comparing results of two different systems consuming a substrait plan.

Yes, that's correct.  However, the eventual plan is to have tests where DuckDB is consuming substrait plans produced by other producers as well.

File: tests/consumers/duckdb_consumer.py
Comment (richtia) [reply tests/consumers/duckdb_consumer.py:22]: Thanks for catching this, i'll update it.

**Comment by pdet** [comment]
LGTM

File: tests/consumers/duckdb_consumer.py
Comment (gforsyth) [inline tests/consumers/duckdb_consumer.py:56]: small nit -- I think we settled on `pathlib` in the initial commit -- do we want to keep that consistent?  I believe you can use the `stem` or `suffix` attributes of a `pathlib.Path` in place of `splitext()`

File: tests/integration/test_duckdb_tpch.py
Comment (gforsyth) [inline tests/integration/test_duckdb_tpch.py:14]:
```suggestion
    Test Class for testing Substrait using DuckDB as a consumer.
```

File: tests/integration/test_duckdb_tpch.py
Comment (gforsyth) [inline tests/integration/test_duckdb_tpch.py:59]: Since you have this utility, maybe it should be rolled into the `consumer` class?  That way you only have to pass in the `file_names` (and then have the `consumer` class figure out the paths using your helper method)

File: tests/consumers/duckdb_consumer.py
Comment (gforsyth) [inline tests/consumers/duckdb_consumer.py:33]: Both of these methods need the `db_connection` -- it could be kept with the `DuckDBConsumer` class and shoved into an attribute so it isn't required to be passed to both static methods (which would no longer be static)

**Review by gforsyth — CHANGES_REQUESTED** [review]
A few small nits and a few questions, nothing major.  Otherwise looks good!

File: tests/consumers/duckdb_consumer.py
Comment (richtia) [reply tests/consumers/duckdb_consumer.py]: Yeah, we did.  Just updated!

File: tests/integration/test_duckdb_tpch.py
Comment (richtia) [reply tests/integration/test_duckdb_tpch.py]: Good idea.  Done.

File: tests/consumers/duckdb_consumer.py
Comment (richtia) [reply tests/consumers/duckdb_consumer.py]: Thanks for the suggestion, this is cleaner!

**Review by gforsyth — APPROVED** [review]
Nice!  Looks good to me.

File: tests/integration/test_duckdb_tpch.py
Comment (westonpace) [inline tests/integration/test_duckdb_tpch.py:27]: What is this yield here for?

**Review by westonpace — COMMENTED** [review]
I'm a little confused what the goal is here.  For Acero, our test corpus consists of `SQL` and `Substrait Query` and we have:

Test: Substrait Query -> Acero -> Arrow
Oracle: SQL -> DuckDb -> Arrow

However, in this PR, we have:

Test: SQL -> DuckDb -> Substrait Query -> DuckDb -> Arrow
Oracle: SQL -> DuckDb -> Arrow

In other words, DuckDb is not running against the same Substrait Query that other consumers (e.g. Acero) are running against.  The `substrait_query` parameter to `test_substrait_query` is completely ignored.

**Comment by richtia** [comment]
>

**Comment by richtia** [comment]
> I'm a little confused what the goal is here. For Acero, our test corpus consists of `SQL` and `Substrait Query` and we have:
> 
> Test: Substrait Query -> Acero -> Arrow Oracle: SQL -> DuckDb -> Arrow
> 
> However, in this PR, we have:
> 
> Test: SQL -> DuckDb -> Substrait Query -> DuckDb -> Arrow Oracle: SQL -> DuckDb -> Arrow
> 
> In other words, DuckDb is not running against the same Substrait Query that other consumers (e.g. Acero) are running against. The `substrait_query` parameter to `test_substrait_query` is completely ignored.

Ahh...I see your confusion.  I should have been more clear.

The reason for the disconnect right now is because the manually generated substrait queries in the Acero tests are being executed against local parquet files, which DuckDB doesn't support yet (I've opened an issue for them to add support here: https://github.com/duckdblabs/substrait/issues/3).  
Whereas DuckDB currently only supports producing/running substrait plans that use named tables (which Acero doesn't support yet.  Vibhatha has the PR in review: https://github.com/apache/arrow/pull/14024/).  

Once there is the same provider support in Acero/DuckDB, I'll open another PR to add the other tests for each consumer so that there is parity in the testing.

The plan is to have: 
Acero Consumer Tests:
1. Isthmus Provider: Substrait Query (local parquet)-> `Acero` -> Arrow (currently supported)
2. DuckDb Provider: SQL -> DuckDb -> Substrait Query (named tables) -> `Acero` -> Arrow (currently not supported)

DuckDB Consumer Tests:
1. Isthmus Provider (using local parquet): Substrait Query (local parquet) -> `DuckDB` -> Arrow (currently not supported)
2. DuckDb Provider (using named tables): SQL -> DuckDb -> Substrait Query (named tables) -> `DuckDB` -> Arrow (currently supported)

I was going to updated the README once the support is all there...but I can also just update with the existing support in this PR if that helps avoid any confusion too.

**Comment by richtia** [comment]
Accidentally closed the PR...reopening

File: tests/integration/test_duckdb_tpch.py
Comment (richtia) [reply tests/integration/test_duckdb_tpch.py:27]: the yield is for the `setup_teardown_class` fixture.

Once all the tests are run, anything after this yield is called.  In this specific case, we close the duckdb connection at the end.

**Comment by westonpace** [comment]
> Once there is the same provider support in Acero/DuckDB, I'll open another PR to add the other tests for each consumer so that there is parity in the testing.

This seems like a reasonable approach.

**Comment by jacques-n** [comment]
It's seems based on comments that this patch doesn't integrates with neither acero nor isthmus. I think it should integrate with at least one before being merged. It sounds like thst would be isthmus given the state of things?

**Comment by westonpace** [comment]
> It's seems based on comments that this patch doesn't integrates with neither acero nor isthmus. I think it should integrate with at least one before being merged. It sounds like thst would be isthmus given the state of things?

I'm not entirely sure that is needed.  It seems like we have four tasks planned.

| Id | Producer | Consumer |
| --- | --- | --- |
| A | Isthmus | Acero |
| B | DuckDb | Acero |
| C | Isthmus | DuckDb |
| D | DuckDb | DuckDb |

Task A has already merged.
Task B is blocked by Acero work.
Task C is blocked by DuckDb work.
Task D is added by this PR.

**Comment by jacques-n** [comment]
Why would D be done here and not in duckdb? It sounds like an internal test of duckdb, not an integration test of two different systems.

**Comment by richtia** [comment]
> Why would D be done here and not in duckdb? It sounds like an internal test of duckdb, not an integration test of two different systems.

D was done mostly as an example to demonstrate how we should use duckdb as a producer or consumer in an integration test with other producers/consumers.  It's modular enough so that we can swap out consumers/producers relatively easily. I think it would be good to keep these here until we have other consumer/producer integration tests with the duckdb producer/consumer.

**Comment by jacques-n** [comment]
> D was done mostly as an example to demonstrate how we should use duckdb as a producer or consumer in an integration test with other producers/consumers. It's modular enough so that we can swap out consumers/producers relatively easily. I think it would be good to keep these here until we have other consumer/producer integration tests with the duckdb producer/consumer.

Examples are super useful. However, I'm not convinced this is the place for examples. How about we put it on an example page on the Substrait website or in a Substrait examples repo. 

Integration testing is sacred from my pov. It's what proves that this isn't a single-system standard. Having a single system talking to itself that may be entirely inconsistent from other systems talking to each other in the same repo feels like it sends the wrong message. It suggests that we're doing integration testing with DuckDB when we aren't (yet).

Given both your pushback, it seems like this is more controversial than I would expect it to be. Maybe we had different intentions for this repo or I'm missing part of the picture that you both have in your heads?

**Comment by westonpace** [comment]
I think the idea is that there are P different producers and Q different queries.  For each consumer you will have `P * Q` different test cases:

```
q_1 x p_1
q_1 x p_2
...
q_1 x p_P
q_2 x p_1
...
q_Q x p_P
```

Eventually that won't be exact because some queries will only make sense for a subset of producers.  Also, today all queries are SQL but I think there may be a future where that isn't the case.

However, the point is that, for each query, a producer should return the correct answer, regardless of the producer that is converting query to plan.  You could imagine a support table:

| Consumer | Isthmus | DuckDb | Ibis | ... | dplyr |
| --- | --- | --- | --- | --- | --- |
| Acero | :heavy_check_mark: | :x: | :heavy_check_mark:  | ... | :heavy_check_mark: |
| DuckDb | :heavy_check_mark: | :heavy_check_mark: | :x: | ... | :heavy_check_mark: |

DuckDb is a producer, so there are a set of test cases created from it:

```
q_1 x DuckDb
q_2 x DuckDb
...
q_Q x DuckDb
```

We could call these tests "Integration" and skip them.  Yielding:

| Consumer | Isthmus | DuckDb | Ibis | ... | dplyr |
| --- | --- | --- | --- | --- | --- |
| Acero | :heavy_check_mark: | :x: | :heavy_check_mark:  | ... | :heavy_check_mark: |
| DuckDb | :heavy_check_mark: | (NA) | :x: | ... | :heavy_check_mark: |

...however it's not clear to me that we have to skip these cases.  If it isn't too much trouble to include them then it seems it will keep things simpler rather than confuse users why there might be an NA in the table.

**Comment by westonpace** [comment]
"Support table" is maybe the wrong word.  "Test result table" might be more descriptive.  In theory a row should have check marks across the board or X's across the board.  In practice we've seen that each producer's nuances tend to lead to scattered X's.  However, this is not something we should be advertising or telling users, which is why "support table" is not really the right word.

**Comment by westonpace** [comment]
> It suggests that we're doing integration testing with DuckDB when we aren't (yet).

I agree that adding this PR is not completely satisfying "add duckdb consumer and tpch tests".  That title would imply to me that we are fully testing duckdb against all possible producers.

Perhaps we can reword the PR title to "add duckdb consumer and tpch tests for duckdb producer" and then make sure we have github issues for "add duckdb consumer and tpch tests for Isthmus producer" and "add acero consumer and tpch tests for duckdb producer".

**Comment by jacques-n** [comment]
I don't like merging something that may have a completely disjunct perspective on what a Substrait plan is. At least early on, let's only merge things here that also have a "trust chain". Once duckdb is in the trust chain via acero and/or isthmus, we can also add the inside-duck connection.

**Comment by pdet** [comment]
Hey guys, 

so for a bit of context, what tpc-h queries can Acero and Isthmus round-trip?

From what I can see in the repo, acero is a consumer-only engine, but I'm not exactly sure how to consume/produce substrait plans from isthmus (I also haven't heard about isthmus before, so any ref pointers would be nice).

If these systems are (or are close to) a ground truth of accepted substrait plans, I'm of course happy to add them to our own substrait CI (We already have a bit on Ibis), since the acero consumption code is in this repo, having it in our CI and going from there should be pretty straightforward, and I can start working on adapting to it. :-).

**Comment by richtia** [comment]
> Hey guys,
> 
> so for a bit of context, what tpc-h queries can Acero and Isthmus round-trip?
> 
> From what I can see in the repo, acero is a consumer-only engine, but I'm not exactly sure how to consume/produce substrait plans from isthmus (I also haven't heard about isthmus before, so any ref pointers would be nice).
> 
> If these systems are (or are close to) a ground truth of accepted substrait plans, I'm of course happy to add them to our own substrait CI (We already have a bit on Ibis), since the acero consumption code is in this repo, having it in our CI and going from there should be pretty straightforward, and I can start working on adapting to it. :-).

Hey Pedro, here's the repo with Isthmus: https://github.com/substrait-io/substrait-java/tree/main/isthmus

The TPC-H substrait plans in this repo were generated using Isthmus and then modified to read data from parquet files since that's is what Acero currently support:
https://github.com/substrait-io/consumer-testing/tree/main/tests/integration/queries/tpch_substrait_plans
(I know there is a plan for duckdb to add that support: https://github.com/duckdblabs/substrait/issues/3).  

By default Isthmus has named tables in the substrait plans though, similar to duckdb produced substrait plans, so you may be able to run those already.

**Comment by westonpace** [comment]
Another good idea might be to run the DuckDb (or whatever producer we are testing) plans through the validator.

**Comment by pdet** [comment]
Hi @westonpace, we already do that in the DuckDB CI, all tpc-h queries go through the validator.

Feel free to adapt the code https://github.com/duckdblabs/substrait/blob/main/test/python/test_validator.py

**Comment by richtia** [comment]
Thanks @pdet
Just added a test to check the validity of substrait plans generated via isthmus as well as duckdb: 
https://github.com/substrait-io/consumer-testing/pull/3/files#diff-41f46f832fcb78dd7d4458cc161defc9bf5e50da0e9e0f570e868566f0b43d69

Looks like Isthmus only need warnings to be overridden, but duckdb needs a few errors to be overridden to pass the validator.

**Pull request closed without merging**
